### PR TITLE
Add --add-host for docker build

### PR DESCRIPTION
**Pull request by tabboud**
Signed-off-by: Tony Abboud <tdabboud@hotmail.com>

**- What I did**
Added the `--add-host` flag to the `docker build` command. This flag allows you to add extra hosts to a containers /etc/hosts file upon building an image.

**- How I did it**
I added another flag to the build handler and added server side code to accept and set the extra hosts when the `/build` endpoint is hit.

**- How to verify it**
I have added an integration test `TestBuildWithExtraHost` which adds hosts whose ip is set to the localhost, but then it tries to ping based off of the hostname that was set.

**- Description for the changelog**
Add the --add-host flag to docker build to add entries to a containers /etc/hosts file when building an image

Implements #30096

### CODE REVIEW

**Comment by LK4D4** [comment]
Can't this be achieved with ARG or whatever?
ping @docker/core-engine-maintainers

**Review by vdemeester — COMMENTED** [review]
ping @tiborvass @tonistiigi @duglin @vieux

**Comment by duglin** [comment]
@LK4D4 can you elaborate? Are you imagining some kind of "cat $ARG >> /etc/hosts" type of thing appearing in the Dockerfile?

File: cli/command/image/build.go [hunk 1]
```diff
@@ -107,6 +109,7 @@ func NewBuildCommand(dockerCli *command.DockerCli) *cobra.Command {
 	flags.BoolVar(&options.compress, "compress", false, "Compress the build context using gzip")
 	flags.StringSliceVar(&options.securityOpt, "security-opt", []string{}, "Security options")
 	flags.StringVar(&options.networkMode, "network", "default", "Set the networking mode for the RUN instructions during build")
+	flags.Var(&options.extraHosts, "add-host", "Add a custom host-to-IP mapping (host:ip)")
```
Comment (duglin) [inline cli/command/image/build.go]: thank you for adding the "form" to the help text - I can never remember that stuff.

File: integration-cli/docker_cli_build_test.go [hunk 2]
```diff
@@ -5565,6 +5565,22 @@ func (s *DockerSuite) TestBuildNetContainer(c *check.C) {
...
 }
 
+func (s *DockerSuite) TestBuildWithExtraHost(c *check.C) {
+	testRequires(c, DaemonIsLinux)
+
+	name := "testbuildwithextrahost"
+	buildImageSuccessfully(c, name,
+		withBuildFlags(
+			"--add-host", "foo:127.0.0.1",
+			"--add-host", "bar:127.0.0.1",
```
Comment (duglin) [inline integration-cli/docker_cli_build_test.go]: I know we're in the design phase and this is premature, but if I don't ask now I'll probably forget....
can you add some negative tests? For example `--add-host myhost:` or `--add-host myhost` and ones with bad IPs. And ones with ipv6

**Comment by duglin** [comment]
In your testing, does the /etc/hosts file get saved in the resulting image?

**Comment by duglin** [comment]
as long as the changes are not saved in the resulting image, this SGTM

File: cli/command/image/build.go (hunk 1 above)
Comment (TDAbboud) [inline cli/command/image/build.go]: 👍 me neither

File: integration-cli/docker_cli_build_test.go (hunk 2 above)
Comment (TDAbboud) [inline integration-cli/docker_cli_build_test.go]: That makes total sense. I was thinking about this when I wrote that test. I will add some more to this PR and push it again.

**Comment by TDAbboud** [comment]
The `/etc/hosts` file is not saved in the resulting image. I actually expressed concern about this in #30096

**Comment by thaJeztah** [comment]
We discussed this in the maintainers meeting, and although we still think setting up a DNS for these purposes, we already implemented `--network` for `docker build`, so let's go ahead and do the same for `--add-host`

**Comment by vdemeester** [comment]
@TDAbboud  needs a rebase 👼

**Comment by TDAbboud** [comment]
@vdemeester I just re-based and pushed the branch again. I also added another test for some invalid extra hosts. 
There currently is no test for a valid IPv6 address, as I was not sure how exactly to implement this. The default `/etc/hosts` file already contains entries for localhost i.e. `::1  localhost` so adding another entry will not work. I think it would be good to test this case as well though, if anyone has a good idea on how to test it.

File: docs/api/version-history.md [hunk 3]
```diff
@@ -34,6 +34,7 @@ keywords: "API, Docker, rcli, REST, documentation"
 * The API version is now required in all API calls. Instead of just requesting, for example, the URL `/containers/json`, you must now request `/v1.25/containers/json`.
 * `GET /version` now returns `MinAPIVersion`.
 * `POST /build` accepts `networkmode` parameter to specify network used during build.
+* `POST /build` now accepts `extrahosts` parameter to specify a host to ip mapping to use during the build.
```
Comment (vdemeester) [inline docs/api/version-history.md]: This is not at the right place, should be in `v1.27 API changes` :angel:

**Review by vdemeester — APPROVED** [review]
Small nit on the API changes, but otherwise LGTM 🦁

File: docs/api/version-history.md (hunk 3 above)
Comment (TDAbboud) [inline docs/api/version-history.md]: Good catch! I just fixed this and re-pushed

**Review by cpuguy83 — APPROVED** [review]
LGTM

We are currently having issues with CI, so we need to make sure to wait for CI to run against this PR.

Thanks!

**Comment by TDAbboud** [comment]
Just pinging to verify all tests have passed and the branch has been rebased

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!

**Comment by arehmandev** [comment]
Can we confirm this is working? I have just recently tried:

 docker build --add-host=docker:10.180.0.1 -t abs
docker exec -it "containerid" sh
cat /etc/hosts - it doesn't appear there

**Comment by thaJeztah** [comment]
the host that's  added with this flag is only used during the build; it deliberately should not persist in the image

**Comment by arehmandev** [comment]
Is this really as intended? I mean other than at runtime, how can you persist host entries to the image? 

- perhaps the best way to approach this would be something like "HOSTENTRY" in the Dockerfile

**Comment by cpuguy83** [comment]
@arehmandev You don't. These are generally host specific settings and are not persisted into the image.

**Comment by arehmandev** [comment]
So I guess its up to the container orchestration system now to plug in these values. Cool thanks for clarification

**Pull request merged**